File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
        
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(node,lower = float('-inf'),upper = float('inf')):
            if not node:
                return True
            if node.val <= lower or node.val >= upper:
                return False
            if not helper(node.right,node.val,upper) or not helper(node.left,lower,node.val): 
                return False
           
            
            return True
        
        
        return(helper(root))

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> res = new ArrayList<Integer>();

    public void inorder(TreeNode root){
        if(root != null){
            // Start exploring the left child
            inorder(root.left);
            res.add(root.val);
            // Start exploring the right child
            inorder(root.right);
        }
    }

    public boolean isValidBST(TreeNode root) {
        inorder(root);

	// Check if the list is sorted or not
	    for(int i = 0; i < res.size()-1; i++){
		// At any point if we find a breach, then we know it is not in a sorted order
		if(res.get(i) >= res.get(i+1)){
			return false;
            }
        }
        
        return true;
}
}